Add SensorPush Cloud integration

sstallion opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds cloud integration for SensorPush devices. It communicates with the publicly available [Cloud API](https://www.sensorpush.com/gateway-cloud-api) using the [sensorpush-api](https://pypi.org/project/sensorpush-api/) and [sensorpush-ha](https://pypi.org/project/sensorpush-ha/) Python packages. Care was taken to ensure that presented devices appeared the same as those created by the existing SensorPush integration. A [G1 WiFi Gateway](https://www.sensorpush.com/products/p/g1-gateway) is required to make use of the Cloud API.

> [!NOTE]
> This PR is a re-submission of https://github.com/home-assistant/core/pull/121890, which fixes several issues reported by reviewers. I would like to send a heartfelt apology to @joostlek, @edenhaus, and @frenck - I was caught between dueling schedules and I wasn't able to do the work needed to get it over the finish line. I was finally able to put in the time over the holiday break to make requested changes in addition to improved unit tests, strict typing, async dependencies, and websession injection (whew!)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36600

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.
1. https://github.com/home-assistant/core/pull/134200
2. https://github.com/home-assistant/core/pull/134196

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

zweckj commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+            email, password = user_input[CONF_EMAIL], user_input[CONF_PASSWORD]
+            await self.async_set_unique_id(email, raise_on_progress=False)
+            self._abort_if_unique_id_configured()
+            try:
+                api = SensorPushCloudApi(self.hass, email, password)
```
```suggestion
            api = SensorPushCloudApi(self.hass, email, password)
            try: 
```
only have code in try that can fail

zweckj commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+class SensorPushCloudConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for SensorPush Cloud."""
+
+    VERSION = 1
```
```suggestion
```
is the default

zweckj commented on homeassistant/components/sensorpush_cloud/const.py:
```diff
@@ -0,0 +1,21 @@
+ATTR_SIGNAL_STRENGTH: Final = "signal_strength"
+ATTR_VAPOR_PRESSURE: Final = "vapor_pressure"
+
+UPDATE_INTERVAL: Final = timedelta(seconds=60)
```
this might be personal preference: if consts are only used in a single file, i'd move them there

zweckj commented on homeassistant/components/sensorpush_cloud/coordinator.py:
```diff
@@ -0,0 +1,39 @@
+class SensorPushCloudCoordinator(DataUpdateCoordinator[dict[str, SensorPushCloudData]]):
+    """SensorPush Cloud coordinator."""
+
+    api: SensorPushCloudApi
```
```suggestion
```
you have it in init already

zweckj commented on homeassistant/components/sensorpush_cloud/coordinator.py:
```diff
@@ -0,0 +1,39 @@
+        email, password = entry.data[CONF_EMAIL], entry.data[CONF_PASSWORD]
+        api = SensorPushCloudApi(hass, email, password)
+        self.helper = SensorPushCloudHelper(api)
+        super().__init__(
+            hass, LOGGER, name=entry.title, update_interval=UPDATE_INTERVAL
+        )
```
```suggestion
        super().__init__(
            hass, LOGGER, name=entry.title, update_interval=UPDATE_INTERVAL, config_entry=entry,
        )
```
and preference. move this first in init

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,162 @@
+) -> None:
+    """Set up SensorPush Cloud sensors."""
+    coordinator: SensorPushCloudCoordinator = entry.runtime_data
+    device_ids = await coordinator.async_get_device_ids()
```
I'd rather do that in `_async_setup` of the coordinator (which gets called during `async_config_entry_first_refresh`) and store it in the coordinator, so we don't have to do IO again

sstallion replied:
The thought at the time was to handle configuring additional entities without requiring a restart. The list returned by the API isn't fixed and is populated based on actions taken on the mobile app.

zweckj replied:
That'll require a bit more work, since this is only called once. You'll need to call this during a regular update and register callback listeners to add the new entities. 
You can check withings binary sensors or tedee for different approaches how to achieve this. I would leave this for a follow up PR though.

sstallion replied:
Thanks! That sounds like a good opportunity to add support for a reconfigure step too. Would it make sense to keep this as-is for now, or should I rework this into `_async_setup`?

zweckj replied:
If you’re aiming for dynamic entities just move this to `_async_update` already

zweckj commented on homeassistant/components/sensorpush_cloud/coordinator.py:
```diff
@@ -0,0 +1,39 @@
+
+    async def _async_update_data(self) -> dict[str, SensorPushCloudData]:
+        """Fetch data from API endpoints."""
+        return await self.helper.async_get_data()
```
can this fail? If yes, re-raise an `UpdateFailed`

sstallion replied:
No failures here - sensorpush-ha will always return a value and log any unexpected exceptions encountered.

zweckj replied:
my suggestion would be to catch exceptions here, so you can raise `UpdateFailure` and let HA handle the retries . The problem with catching upstream is, for example, how would you know that authentication expired/got invalid (in which case we normally raise `ConfigEntryAuthFailed` to trigger a reauth flow, but that's for a later PR anyway)

sstallion replied:
Sounds good. I've moved exception handling outside of the call with a call to `LOGGER.exception` for now. I don't expect this to occur since the API library will handle retries before giving up and is not something I've seen in testing for the last few months. Once support for translatable error messages goes in, this can be converted to throw a `SensorPushCloudUpdateError` or similar instead.

zweckj reviewed (changes requested):
Thanks for putting in the effort and going for a second attempt at this! However, I see the libraries are not compliant yet as, `sensorpush-api`'s source code is not public and  under an OSI approved license https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/dependency-transparency

sstallion commented:
> Thanks for putting in the effort and going for a second attempt at this! However, I see the libraries are not compliant yet as, `sensorpush-api`'s source code is not public and under an OSI approved license https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/dependency-transparency

Thanks! The source code for the Python library is public (I'm the author), however you are correct that the license is not OSI approved. I could use some advice here if you don't mind. OpenAPI generated the original source, which I then modified to work with HA. The license information was generated based on information in the Swagger definition, which references SensorPush's EULA. What have other authors done in this case?

zweckj commented:
I could only find the -ha lib on GitHub. The other one I could only find on PyPi but not in a repo. I guess most will publish to a repo and adapt the license there

sstallion commented:
> I could only find the -ha lib on GitHub. The other one I could only find on PyPi but not in a repo. I guess most will publish to a repo and adapt the license there

Huh, that's odd. You should be able to find both libraries on PyPI:
1. https://pypi.org/project/sensorpush-api/
2. https://pypi.org/project/sensorpush-ha/

I'll work on re-licensing sensorpush-api in the meantime. I suspect this is just an artifact of OpenAPI.

zweckj commented:
Yes on PyPi. But that’s not a code repo (like GitHub). You'll need to put the code (+releases & CICD) there as well.

sstallion commented:
> Yes on PyPi. But that’s not a code repo (like GitHub). You'll need to put the code (+releases & CICD) there as well.

Ah! That's strange, I'm not sure why the repository link is absent. I'll look into it.

Everything is on GitHub and releases are managed by CI/CD:
1. https://github.com/sstallion/sensorpush-api
2. https://github.com/sstallion/sensorpush-ha

sstallion commented:
Hi @zweckj - I've corrected the license terms for `sensorpush-api` and made a new release. All dependencies use OSI-approved licenses and releases are managed by CI/CD. The PyPI page has also been updated with the correct metadata. Thanks for catching this!

sstallion commented:
I've sanded down a few more edges and made another pass at cleanup. The PR should be ready for review. Thanks in advance!

zweckj commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,48 @@
+            try:
+                await api.async_authorize()
+            except SensorPushCloudError as e:
+                errors["base"] = str(e)
```
we want those translateable, so set it to a key which is in your `strings.json` (that could mean you'll need more cases of course)

sstallion replied:
This might be tough to tackle in the first PR. The Cloud API only provides a free-form message field for errors without corresponding status codes. At the moment, there does not seem to be a way to distinguish between a temporary service outage versus other errors (eg. authentication) without matching this message string, which could be brittle. Would it be possible to defer this until a better solution can be found?

zweckj replied:
Yes, let it fall back to `unknown` for the UI and error log the whole error

sstallion replied:
I believe this is the current behavior. `SensorPushCloudError` will contain the API-generated / UI error message. Any other exception will fall back to `unknown`.

zweckj replied:
But you set `errors[base]` to the actual error message. This key is will be used to search for a translateable error message in your strings.json, so we don’t have something to translate here

sstallion replied:
Gotcha. Is there a better way to display a user-friendly error message that is not yet translated?

sstallion replied:
Example behavior:
![Screenshot 2025-01-02 at 14-40-34 Settings – Home Assistant](https://github.com/user-attachments/assets/744f89a3-3269-40da-b5b0-a54ad9186095)

sstallion replied:
I'm not seeing much in core for this. I suspect this works by accident since there is no key that matches the error message. I'll add a generic connection error for now until translatable errors can be added in a subsequent PR.

zweckj commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,48 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_EMAIL): str,
+                    vol.Required(CONF_PASSWORD): str,
```
if you use `TextSelector`s here you can support autofill (https://developers.home-assistant.io/docs/data_entry_flow_index/#enabling-browser-autofill)

zweckj commented on homeassistant/components/sensorpush_cloud/strings.json:
```diff
@@ -0,0 +1,35 @@
+        "data": {
+          "email": "[%key:common::config_flow::data::email%]",
+          "password": "[%key:common::config_flow::data::password%]"
+        }
```
add data_descriptions for those

zweckj commented on homeassistant/components/sensorpush_cloud/coordinator.py:
```diff
@@ -0,0 +1,35 @@
+            config_entry=entry,
+        )
+        email, password = entry.data[CONF_EMAIL], entry.data[CONF_PASSWORD]
+        api = SensorPushCloudApi(hass, email, password)
```
I didn't look into the library yet, but passing the `hass` object to it rings a bit of an alarm bell. Because you'll couple your library to HA objects, which is not good design wise, plus it's normally unnecessary

sstallion replied:
I completely agree. I think this was an artifact of moving `api.py` out of the integration and into a new package. I'll get this cleaned up ASAP.

zweckj commented on homeassistant/components/sensorpush_cloud/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/sensorpush_cloud",
+  "iot_class": "cloud_polling",
+  "loggers": ["sensorpush_api", "sensorpush_ha"],
```
```suggestion
  "loggers": ["sensorpush_api", "sensorpush_ha"],
  "quality_scale": "bronze",
```

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  unique-config-entry: done
+
+  # Silver
+  action-exceptions: done
```
exempt, no custom actions and only read-only platform

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+rules:
+  # Bronze
+  action-setup: done
```
exempt, no custom actions

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions: done
```
exempt, no custom actions

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+
+  # Silver
+  action-exceptions: done
+  config-entry-unloading: todo
```
```suggestion
  config-entry-unloading: done
```

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  # Silver
+  action-exceptions: done
+  config-entry-unloading: todo
+  docs-configuration-parameters: done
```
exempt, no options flow

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  docs-installation-parameters: done
+  entity-unavailable: done
+  integration-owner: done
+  log-when-unavailable: todo
```
done, coordinator handles that for you (if you allow the exceptions to come to it)

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  entity-disabled-by-default: done
+  entity-translations: todo
+  exception-translations: todo
+  icon-translations: todo
```
imo it would be nice to add icons for the 4 entites, that don't get one from the deviceclass

sstallion replied:
Sure thing - which entities were you thinking of? I believe all of the sensors have a device class assigned.

zweckj replied:
That’s no requirement, if the default ones are ok, that’s also fine

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  # Platinum
+  async-dependency: done
+  inject-websession: done
+  strict-typing: done
```
you are missing an entry to the `.strict-typing` file

sstallion replied:
It looks like there is an existing regular expression that should catch this (`homeassistant.components.sensor.*`). Is it preferable to add an explicit pattern?

zweckj replied:
I don't think I understand what you mean? My comment was that integrations that strictly type need to add themselves to `.strict-typing` which will also change the mypy config on commit so mypy checks your integration strictly.

sstallion replied:
Sorry, this may be a misunderstanding on my part. I believe this integration is already checked due to L424 in `.strict-typing`. I'm happy to add a line, but I believe it will be redundant since an existing pattern already matches this component.

sstallion replied:
Apologies, I completely misunderstood the contents of the file. I'll add this ASAP.

zweckj replied:
ah now I getwhat you meant. It's no regex, so you'll need `homeassistant.components.sensorpush_cloud.*` in there

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,156 @@
+
+        device = self.coordinator.data[device_id]
+        self._attr_unique_id = f"{device.device_id}_{entity_description.key}"
+        self._attr_device_info = device.device_info(DOMAIN)
```
I'd rather not build the device info in the library. It has a couple of downsides, one being you'll needing the `DeviceInfo` type in the lib

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,156 @@
+        self.entity_description = entity_description
+        self.device_id = device_id
+
+        if device_id not in self.coordinator.data:
+            LOGGER.warning("Ignoring inactive device: %s", device_id)
+            self._attr_available = False
+            return
```
this can't happen, can it? We are looping through the device_ids in L114

sstallion replied:
Good catch! I think this is leftover from the first PR when device IDs were kept in the config entry. I'll remove this.

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,156 @@
+    @property
+    def available(self) -> bool:
+        """Return true if entity is available."""
+        if self.device_id not in self.coordinator.data:
+            return False  # inactive device
```
if the device can really be missing from `data` you'll need a `get` in L156. `super().available` will make this entity unavaialbe when the update fails (again, given you let the errors come to an `UpdateFailed` exception)

sstallion replied:
Same as above; will remove.

sstallion commented:
I believe I've addressed all of the review feedback so far. The PR should be ready for another round. Thanks again for the detailed feedback - it is much appreciated!

sstallion commented:
Rebased and made minor corrections to quality scale for documentation.

zweckj commented on homeassistant/components/sensorpush_cloud/coordinator.py:
```diff
@@ -0,0 +1,47 @@
+        try:
+            data = await self.helper.async_get_data()
+        except SensorPushCloudError:
+            LOGGER.exception("Unexpected error")
```
I didn't look into you lib yet, but this probably still needs work. On any exception we should raise `UpdateFailed` here, so that the entities can react on that.

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,155 @@
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return true if entity is available."""
+        last_update = self.coordinator.data[self.device_id].last_update
+        return dt_util.utcnow() < (last_update + MAX_TIME_BETWEEN_UPDATES)
```
this will also get easier with the `UpdateFailed` exception

sstallion replied:
SensorPush's cloud API does not indicate an error if a sensor is no longer available, it just continues to post the latest result. I'll definitely add support for `UpdateFailed`, but this may need to stay to properly show when sensors are unavailable.

joostlek replied:
I think we should also incorporate `super().available` into this mix, this way if your request fails, it will go unavailable.

Also, should we make the entity unavailable if device_id is no longer in self.coordinator.data?

zweckj commented on tests/components/sensorpush_cloud/test_sensor.py:
```diff
@@ -0,0 +1,58 @@
+
+@pytest.mark.data(MOCK_DATA)
+@pytest.mark.usefixtures("entity_registry_enabled_by_default")
+async def test_create_sensors(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+    mock_helper: AsyncMock,
+) -> None:
+    """Test we can create sensors."""
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    for entity_id in MOCK_ENTITY_IDS:
+        entity = hass.states.get(entity_id)
+        assert entity
+        assert entity.state != STATE_UNAVAILABLE
```
just use the `snapshot_platform` helper for this entire test.

zweckj commented on tests/components/sensorpush_cloud/test_coordinator.py:
we don't tests internals (coordinator) directly, we only test the effect it has on the entities

zweckj commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,92 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
compare `enums` with `is` (the rest below as well)
```suggestion
    assert result["type"] is FlowResultType.FORM
```

zweckj commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,92 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_api_error(hass: HomeAssistant, mock_api: AsyncMock) -> None:
```
those two tests can be one with parametrization.

zweckj commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,92 @@
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "user"
+    assert result["errors"] == {"base": "invalid_auth"}
```
also, for config flow tests, make sure you bring them to a terminal state (create_entry/abort). That means in this case, fix the errors now and finish the test to show we can recover from errors.

sstallion commented:
Latest round of changes in! PR is ready for review - thanks again for the feedback!

joostlek commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            email, password = user_input[CONF_EMAIL], user_input[CONF_PASSWORD]
+            await self.async_set_unique_id(email, raise_on_progress=False)
```
```suggestion
            await self.async_set_unique_id(email)
```
Raise on progress only makes sense if you have a discovery flow

joostlek commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            email, password = user_input[CONF_EMAIL], user_input[CONF_PASSWORD]
+            await self.async_set_unique_id(email, raise_on_progress=False)
```
also, do you get any tokens from sensorpush? Are we sure we don't have an user id to use

sstallion replied:
The only token we receive is a short-lived token used for API calls, nothing that would be suitable as a unique ID. SensorPush also relies on email addresses rather than user IDs.

joostlek replied:
Are they JWTs by any chance?

sstallion replied:
No, it appears to be completely bespoke.

joostlek commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,155 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up SensorPush Cloud sensors."""
+    coordinator: SensorPushCloudCoordinator = entry.runtime_data
```
```suggestion
    coordinator = entry.runtime_data
```

joostlek commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,155 @@
+        self.entity_description = entity_description
+        self.device_id = device_id
+
+        device = self.coordinator.data[device_id]
```
```suggestion
        device = coordinator.data[device_id]
```

joostlek commented on homeassistant/components/sensorpush_cloud/strings.json:
```diff
@@ -0,0 +1,40 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Note: To activate API access, log in to the [Gateway Cloud Dashboard](https://dashboard.sensorpush.com/) and agree to the terms of service. Devices are not available until activated with the SensorPush app on iOS or Android.",
```
```suggestion
        "description": "To activate API access, log in to the [Gateway Cloud Dashboard](https://dashboard.sensorpush.com/) and agree to the terms of service. Devices are not available until activated with the SensorPush app on iOS or Android.",
```

joostlek commented on tests/components/sensorpush_cloud/conftest.py:
```diff
@@ -0,0 +1,83 @@
+from tests.common import MockConfigEntry
+
+
+def pytest_configure(config: pytest.Config) -> None:
+    """Register markers for tests that use data."""
+    config.addinivalue_line("markers", "data: mark test to run with specified data")
```
whats this?

sstallion replied:
It registers a custom marker in pytest. Without it, you end up with a warning:
```
test_config_flow.py:19
  /Users/sstallion/Source/homeassistant-core/tests/components/sensorpush_cloud/test_config_flow.py:19: PytestUnknownMarkWarning: Unknown pytest.mark.data - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.data(MOCK_DATA)
```
That said, this will be removed once I move MOCK_DATA out of a marker and into a dedicated mock class.

joostlek commented on tests/components/sensorpush_cloud/conftest.py:
```diff
@@ -0,0 +1,83 @@
+        yield mock_helper
+
+
+@pytest.fixture
+def make_config_entry(
+    request: pytest.FixtureRequest,
+) -> Callable[[dict[str, Any] | None], MockConfigEntry]:
+    """ConfigEntry mock factory."""
+
+    def _make_config_entry(data: dict[str, Any] | None = None):
+        default_data = {
+            CONF_EMAIL: "test@example.com",
+            CONF_PASSWORD: "test-password",
+        }
+        data = default_data if data is None else default_data | data
+        return MockConfigEntry(domain=DOMAIN, data=data, entry_id=data[CONF_EMAIL])
+
+    return _make_config_entry
+
+
+@pytest.fixture
+def mock_config_entry(
+    make_config_entry: Callable[[dict[str, Any] | None], MockConfigEntry],
+) -> MockConfigEntry:
+    """ConfigEntry mock."""
+    return make_config_entry()
```
We only have 2 fields in the config entry data. IMO it does not give a reason to make such logic. Like if you had 7 fields and 3 nested fields to support 10 different devices, I would understand why we need it, but in this case we only have 2 fields which are not of relevance to the devices. So I would make this KISS

sstallion replied:
Would you prefer that the fixture be simpler or that `MockConfigEntry` is instantiated in place?

joostlek replied:
You can just have a fixture that returns a MockConfigEntry and that's it

joostlek commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+) -> None:
+    """Test we get the form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

joostlek commented on tests/components/sensorpush_cloud/conftest.py:
```diff
@@ -0,0 +1,83 @@
+    config.addinivalue_line("markers", "data: mark test to run with specified data")
+
+
+@pytest.fixture
+def mock_api(request: pytest.FixtureRequest) -> Generator[AsyncMock]:
+    """Override SensorPushCloudApi."""
+    mock_api = AsyncMock(SensorPushCloudApi)
+    with (
+        patch(
+            "homeassistant.components.sensorpush_cloud.config_flow.SensorPushCloudApi",
+            return_value=mock_api,
+        ),
+    ):
+        yield mock_api
+
+
+@pytest.fixture
+def mock_helper(request: pytest.FixtureRequest) -> Generator[AsyncMock]:
+    """Override SensorPushCloudHelper."""
+    mock_helper = AsyncMock(SensorPushCloudHelper)
+    marker = request.node.get_closest_marker("data")
+    if marker is not None:
+        mock_helper.async_get_data.return_value = marker.args[0]
+    with (
+        patch(
+            "homeassistant.components.sensorpush_cloud.coordinator.SensorPushCloudHelper",
+            return_value=mock_helper,
+        ),
+    ):
+        yield mock_helper
```
If you're only going to be returning the same MOCK_DATA, I would just suggest to just build that into the mock instead of building a whole fixture request framework for it

joostlek commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+            CONF_PASSWORD: "test-password",
+        },
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/sensorpush_cloud/test_config_flow.py:
I am missing a test where we setup the same account twice

joostlek reviewed (changes requested):
Hey welcome back, how have you been?

sstallion commented:
> Hey welcome back, how have you been?

Good! A little less busy this time around! I hope everyone had a good holiday break.

sstallion commented:
@zweckj @joostlek I believe I've incorporated requested changes. Please take a look when you have a moment - thanks again for the detailed feedback!

joostlek commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+from tests.common import MockConfigEntry
+
+
+async def test_form(
+    hass: HomeAssistant,
+    mock_api: AsyncMock,
+    mock_helper: AsyncMock,
+    mock_setup_entry: AsyncMock,
+) -> None:
+    """Test we get the form."""
```
Stale test name and docstring, the test tests more

joostlek commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        result["flow_id"],
+        CONF_DATA,
+    )
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == "test@example.com"
+    assert result["data"] == CONF_DATA
+    assert len(mock_setup_entry.mock_calls) == 1
```
assert unique id

joostlek commented on tests/components/sensorpush_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], CONF_DATA
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

sstallion commented:
Requested changes have been made. Thanks again!

sstallion commented:
Apologies, I just rebased the branch without realizing there was a workflow run!

sstallion commented:
@joostlek Checking in! Is there anything I can do help?

zweckj commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+            except SensorPushCloudError:
+                LOGGER.exception("Invalid authentication")
+                errors["base"] = "invalid_auth"
+            except Exception:  # noqa: BLE001
```
do you really need the `noqa` here?

sstallion replied:
Unfortunately yes, otherwise `ruff` will complain.

zweckj replied:
it shouldn't though as you have a `LOGGER.exception` there
see <https://docs.astral.sh/ruff/rules/blind-except/> at the bottom

sstallion replied:
It may be because `ruff` does not understand that `LOGGER.exception()` is equivalent to `logging.exception()`. FWIW, there are 75 other config flows that contain `noqa: BLE001` declarations. I'm not sure of the best way to go about fixing this - is this something that could be addressed in a subsequent PR?

joostlek replied:
Ruff indeed does not see that the logger is a logger, I already opened <https://github.com/astral-sh/ruff/issues/15473> for that

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  test-coverage: todo
+
+  # Gold
+  devices: todo
```
```suggestion
  devices: done
```

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+    comment: Integration does not register custom actions.
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: done
```
this is missing

sstallion replied:
Thanks for catching this! I've updated the documentation PR to include the standard removal section.

zweckj commented on homeassistant/components/sensorpush_cloud/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  discovery: todo
+  docs-data-update: todo
+  docs-examples: todo
+  docs-known-limitations: done
```
or it is unclear to me
```suggestion
  docs-known-limitations: todo
```

sstallion replied:
There is a small section near the bottom of the documentation that states:
> Note: atmospheric_pressure is not available in HT1 series devices.

This is the only limitation that I am aware of in this integration.

zweckj replied:
can we move that to a "known limitations" section, so it's more standardized.

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        """Return true if entity is available."""
+        if self.device_id in self.coordinator.data:
+            last_update = self.coordinator.data[self.device_id].last_update
+            return dt_util.utcnow() < (last_update + MAX_TIME_BETWEEN_UPDATES)
```
```suggestion
            return dt_util.utcnow() < (last_update + MAX_TIME_BETWEEN_UPDATES) and super().available
```

zweckj reviewed (changes requested):
couple small things from my end. Regarding your question what you can do: Just wait 😉 we'll look at it when we can

sstallion commented:
> couple small things from my end. Regarding your question what you can do: Just wait 😉 we'll look at it when we can

Will do!

sstallion commented:
Thanks again @zweckj! I've incorporated your suggestions, updated the branch, and put the PR up for review.

zweckj commented on tests/components/sensorpush_cloud/conftest.py:
```diff
@@ -0,0 +1,62 @@
+
+
+@pytest.fixture
+def mock_helper(request: pytest.FixtureRequest) -> Generator[AsyncMock]:
+    """Override SensorPushCloudHelper."""
+    mock_helper = AsyncMock(SensorPushCloudHelper)
+    mock_helper.async_get_data.return_value = MOCK_DATA
+    with (
+        patch(
+            "homeassistant.components.sensorpush_cloud.coordinator.SensorPushCloudHelper",
+            return_value=mock_helper,
+        ),
+    ):
+        yield mock_helper
```
```suggestion
def mock_helper(request: pytest.FixtureRequest) -> Generator[MagicMock]:
    """Override SensorPushCloudHelper."""
    with (
        patch(
            "homeassistant.components.sensorpush_cloud.coordinator.SensorPushCloudHelper",
             autospec=True,
        ) as mock_helper,
    ):
    	helper = mock_helper.return_value
    	helper.async_get_data.return_value = MOCK_DATA
        yield helper
```
is the more standard approach imo

sstallion commented:
@zweckj Requested changes have been incorporated into the documentation and this PR. Thanks again for the feedback!

zweckj reviewed (approved):
LGTM, let's see if Joost has some final remarks

sstallion commented:
@joostlek No rush, I just want to keep the PR from going stale.

ZyberSE commented:
@joostlek any chance this PR could make it into the feb release? Would be 🚀

joostlek commented:
Nope, it's a new feature. I wanted to look at it again this week, but the days before beta were too busy to find time

ZyberSE commented:
> Nope, it's a new feature. I wanted to look at it again this week, but the days before beta were too busy to find time

I see. If it's of any help I've been running iterations of this code since last summer with 19 sensors.

sstallion commented:
Hey @joostlek - it looks like I need to update the documentation PR since this isn't being considered for 2025.02. Do you have an idea of which release I should target? Thanks again!

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,158 @@
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: SensorPushCloudConfigEntry,
```
```suggestion
    entry: AddConfigEntryEntitiesCallback,
```

zweckj commented on homeassistant/components/sensorpush_cloud/sensor.py:
```diff
@@ -0,0 +1,158 @@
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
```
```suggestion
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
```
those have changed in the meantime (might need a rebase)

zweckj commented:
>  Do you have an idea of which release I should target? Thanks again!

Dw about that, we'll fix that if needed

sstallion commented:
Thanks @zweckj! I've rebased the branch and pulled in changes from the dev branch. Would it be possible to approve a workflow run for the PR? Thanks again!

joostlek commented on homeassistant/components/sensorpush_cloud/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+            api = SensorPushCloudApi(email, password, clientsession)
+            try:
+                await api.async_authorize()
+            except SensorPushCloudError:
+                LOGGER.exception("Invalid authentication")
```
Can we be more specific in catching exceptions? As in, are all exceptions an authentication error? If we know what the exception is, we should not use `LOGGER.exception` as we already know what was up

sstallion replied:
I could use some advice here. SensorPush's API (usually) provides a human readable string to explain the error, but they have yet to use status codes so we can determine the cause programmatically. So, I have a non-translatable string at my disposal, but nothing else. Is there a way to expose this string to the user for now and try to work with SensorPush in the background to start using status codes?

joostlek replied:
I am not 100% sure, maybe you can use placeholders in errors.

Otherwise, we need to check the common ones and raise proper exceptions. This can just be abstracted away in the library to raise the proper exception when a certain message comes

joostlek commented on tests/components/sensorpush_cloud/conftest.py:
```diff
@@ -0,0 +1,62 @@
+
+
+@pytest.fixture
+def mock_config_entry(
+    request: pytest.FixtureRequest,
+) -> MockConfigEntry:
```
```suggestion
def mock_config_entry() -> MockConfigEntry:
```

joostlek reviewed (commented):
Last things

sstallion commented:
@joostlek Requested changes have been made. I was able to clean up authentication in the library. Thanks again for the detailed review feedback!

sstallion committed:
Add SensorPush Cloud integration

sstallion commented:
@joostlek Circling back around - is there anything else I can do get this ready to go?

joostlek commented:
@sstallion Be sure to send me a message on discord

sstallion commented:
> @sstallion Be sure to send me a message on discord

Will do. Thanks again!
